doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+the event queue. Examples of events might be "new connection", "new stream",
+"stream readable", "stream close", etc. Optionally events can be filtered for
+only a given `SSL` object. If the event queue is empty then a new packet will be
+read from the underlying `BIO`s (if possible) and any resulting events are added
+to the event queue.
+
+Also associated with the `SSL_EVENT_CTX` is a new connection queue and a new
+stream queue. In the case that a "new connection" event is received then the
+application would need to create a new SSL object and then call `SSL_accept()` on
+it. The `SSL` object will have inherited the `SSL_EVENT_CTX` to use from the
+`SSL_CTX` and will inspect the new connection queue and automatically configure
+the `SSL` object for the new incoming connection.
+
+An application can (optionally) insert new connections into the connection
+queue directly by calling `SSL_EVENT_CTX_add_new_connection()`. This would also
+result in a "new connection" event being issued. This would not be necessary for
+QUIC applications because new connections would be automatically added to the
+queue. However it might be useful for (D)TLS applications where only the
+application knows when a new connection has arrived, i.e. OpenSSL cannot detect
+this automatically.
+
+New streams can be initiated by calling `SSL_new_stream()` and passing an
+existing `SSL` object for some other stream from the same connection. Subsequently,
+`SSL_connect()` will need to be called on the new `SSL` object. This could be made
+to work for (D)TLS client applications which could automatically configure the
+SSL object to try a resumption handshake. On the server side this would always
+fail for (D)TLS applications.
+
+New streams can also be initiated by the peer in which case a "new stream"
+event would be issued by the `SSL_EVENT_CTX`. In that case details of the
+incoming stream would be added to the new stream queue inside the `SSL_EVENT_CTX`.
+The application would "accept" the incoming stream by calling
+`SSL_accept_stream()`. This would only occur for QUIC applications - never for
+(D)TLS applications.
+
+A QUIC based application will need to handle various time based events. An
+application can call `SSL_get_next_tick()` to get the next timeout event for a
+given `SSL` object. For a DTLS application this is exactly equivalent to calling
+`DTLSv1_get_timeout()`. Calling `SSL_tick()` on an `SSL` object will enable it
+to process any outstanding time based events. For a DTLS application this is
+exactly equivalent to calling `DTLSv1_handle_timeout()`. In a QUIC application
+only the `SSL` object for a default stream would have time based events. For a TLS
+application there would never be time based events (so calling `SSL_tick()` will
+have no effect).
+
+## Changes to the BIO API
+
+In order to support QUIC the current `BIO` API will need to be extended. The
+existing "dgram" `BIO` can be used for UDP. However this only has very limited
```

Comment:
Devil's advocate question: is reusing the existing "dgram" BIO the best beginning?

I.e. is the unconnected UDP BIO sufficiently different from the connected UDP BIO to justify a second BIO type?  Essentially, is the code sharing worth the differences?

Reply:
I think that would be quite confusing and would probably end up quite a lot of duplicated code. I don't think separating the two is worth it.